### PR TITLE
Completed piece placement

### PR DESCRIPTION
**Pull request by raspberrymilkyway**
TL;DR -- Added Pieces class, added piece placement in frontend and backend

Added Turn class (enum)
Added Piece class (internal use)
Added Pieces class (to be used to assign pieces)
Added GameManager class (backend use - bridge GUI and backend functionality)

Added piece-placement-related Board logic
Added piece-placement-related BoardGUI logic
Added some error checking (though we may want to restructure errors)

Edited BoardGUI init call a bit -- board_size is literally the size of the board, so I added a piece count
Edited BoardGUI imports -- imo, GameManager is the only thing that should be imported; this keeps the backend and frontend initializations and functions separate
Edited Board to have a valid_moves list for the current board
Renamed some Board, Cell, and BoardGUI functions so they use snake_case
Moved GameGUI line so board isn't redrawn from the beginning every time a cell is clicked -- pieces need to persist!

### CODE REVIEW

File: backend/Piece.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    #         pos.append(piece.get_position())
+    #     return pos
+
+    # private methods for increasing piece counts
+    def __increase_count_white_placed(self):
+        self.count_white_placed += 1
+    def __increase_count_black_placed(self):
+        self.count_black_placed += 1
+    
+    #only all_pieces_placed is public!
```
[the remark this answers is not in the document]
Comment (raspberrymilkyway) [reply backend/Piece.py:62]: They are public, but adding two underscores makes a function "private". It's more of a naming standard than an actual language feature -- my bad, I could have made that clearer. The underscores tell the IDE (and any programmer who's aware) that the function should not be used outside of its class. (Depending on the IDE, that could raise a warning or something. It doesn't do anything to compilation or runtime though -- you can still call that function outside of the class.)

Don't know if you'd find the documentation helpful, but here's a section in the Python docs about private variables: [https://docs.python.org/3/tutorial/classes.html#tut-private](url)

File: backend/Board.py [hunk 2]
```diff
@@ -17,12 +17,14 @@ def create_initial_board(self, size):
...
         
         # set the state of the cells to EMPTY
+        self.valid_moves = []
         self.set_valid_cells_to_empty()
         # print(self.board)
 
         return self.board
     
-    def get_valid_moves(self, row):
+    #is this valid movement for 6, 9, and 12?
```
[the remark this answers is not in the document]
Comment (raspberrymilkyway) [reply backend/Board.py:26]: Alright, nice! Yeah, I'll note that we need to change that.

File: backend/Piece.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EXC3ll3NTrhyTHM) [reply backend/Piece.py:62]: Thats interesting! Good to know

File: frontend/BoardGUI.py [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
         self.game = game
         self.board_image = self.get_scaled_image('resources/board/board.png', [WIN_SIZE] * 2)
         self.cell_size = WIN_SIZE // board_size
         self.black_piece_image = self.get_scaled_image('resources/pieces/black_piece.png', [self.cell_size] * 2)
         self.white_piece_image = self.get_scaled_image('resources/pieces/white_piece.png', [self.cell_size] * 2)
-        self.board = Board(board_size)
-        
+        # self.board = Board(board_size)
+        self.backend_game = GameManager(board_size, total_pieces)
+        self.count = 0
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/BoardGUI.py:17]: @raspberrymilkyway what is this variable used for?

File: frontend/BoardGUI.py [hunk 4]
```diff
@@ -31,12 +33,22 @@ def draw_piece(self, piece, position):
...
         current_cell = vec2(pg.mouse.get_pos()) // self.cell_size
         col, row = map(int, current_cell)
         left_click = pg.mouse.get_pressed()[0]
         if left_click:
-            print(self.board.board[row][col].state)
+            # print(self.backend_game.get_board(row,col), self.count)
+            self.count += 1
+
+            if not self.backend_game.placement_complete():
+                x = self.backend_game.piece_placement(row, col)
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/BoardGUI.py:45]: Can we give this a more descriptive name?

File: frontend/BoardGUI.py (hunk 3 above)
Comment (raspberrymilkyway) [reply frontend/BoardGUI.py:17]: Oh, if that's about the count variable, that's supposed to be commented out -- whoops. There's a line somewhere near that (that's commented out) that prints a count of how many times the clicked function has run next to the description of what's in the clicked cell spot. I must've missed commenting out the counter and its initialization.

Comment (raspberrymilkyway) [reply frontend/BoardGUI.py:17]: If it's about the backend_game variable, that's used to handle any backend game management functionality. I can send you an image of the current class interactions if you'd like, but basically it's to separate the backend and the frontend more cleanly -- rather than calling the Board class directly, you'd interact with it through the GameManager class. It also allows the backend to stand and run on its own, which might make testing easier.

File: frontend/GameGUI.py [hunk 5]
```diff
@@ -23,12 +24,13 @@ def check_events(self):
...
     def run_game(self):
+        self.board.build_board() #does this need to be called constantly?
+    
         while True:
             self.check_events()
             pg.display.update()
             self.clock.tick(60)
-            self.board.build_board()
-            self.board.getCellClicked()
+            self.board.get_cell_clicked() #this runs too much. reduce or face errors
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/GameGUI.py:33]: I think this is a good point because when console logging the cell its clicked, it shows it was clicked like 4 or 5 times per click. Not sure what to do about it though. We need an investigation on it.

File: frontend/GameGUI.py [hunk 6]
```diff
@@ -23,12 +24,13 @@ def check_events(self):
                 sys.exit()
 
     def run_game(self):
+        self.board.build_board() #does this need to be called constantly?
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/GameGUI.py:27]: Its only called once right? At the beginning of the program?

Comment (EXC3ll3NTrhyTHM) [reply frontend/GameGUI.py:27]: Its outside the while loop

File: frontend/BoardGUI.py (hunk 4 above)
Comment (raspberrymilkyway) [reply frontend/BoardGUI.py:45]: Sure -- is "flag" descriptive enough or do you want a more specific name? It should only be used directly afterwards.

I could also move it directly to the if-statement and not store the return value at all, if you think that'd read better.

File: frontend/GameGUI.py (hunk 5 above)
Comment (EXC3ll3NTrhyTHM) [reply frontend/GameGUI.py:33]: Investigation issue logged https://github.com/Advanced-Software-Engineering-UMKC/Nine-Mens-Morris/issues/21

Comment (raspberrymilkyway) [reply frontend/GameGUI.py:33]: Yeah, I don't know how to fix that, but I did think it was important to note. The number of times it runs depends entirely on the length of the click -- if you hold the left mouse button down, it'll go on infinitely, lol. Makes testing the piece placement pretty easy though, since you can just drag the mouse around the screen.

File: frontend/GameGUI.py (hunk 6 above)
Comment (raspberrymilkyway) [reply frontend/GameGUI.py:27]: It is now -- that was a comment I forgot to remove when I added the pull request, since that line was initially in the while loop. I removed it from my local branch immediately after, but that doesn't reflect here. My bad!

Comment (raspberrymilkyway) [reply frontend/GameGUI.py:27]: I'll remove it when I push the image/example README change.

File: frontend/BoardGUI.py (hunk 4 above)
Comment (raspberrymilkyway) [reply frontend/BoardGUI.py:45]: I'm going to change it to "flag" when I push the image/example README change, but we can always swap it later if needed.

File: frontend/BoardGUI.py [hunk 7]
```diff
@@ -31,12 +33,22 @@ def draw_piece(self, piece, position):
...
+                # print(x, "\n")
+                if x == 1:
+                    self.draw_piece(self.backend_game.get_turn(),current_cell * self.cell_size)
+                    self.backend_game.end_turn()
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/BoardGUI.py:49]: We should break out these 2 lines of the function as it imposes on separation of concerns. These line could be in a condition for if `get_cell_clicked` returns a cell. It coud be at this place in the code https://github.com/Advanced-Software-Engineering-UMKC/Nine-Mens-Morris/blob/main/frontend/GameGUI.py#L29-L33. There may be a better way to do it as well.

File: frontend/BoardGUI.py (hunk 3 above)
Comment (raspberrymilkyway) [reply frontend/BoardGUI.py:17]: I'll comment out the count variable and its call when I push the image/example README change.

File: frontend/BoardGUI.py [hunk 8]
```diff
@@ -31,12 +33,22 @@ def draw_piece(self, piece, position):
...
         if left_click:
-            print(self.board.board[row][col].state)
+            # print(self.backend_game.get_board(row,col), self.count)
+            self.count += 1
+
+            if not self.backend_game.placement_complete():
+                x = self.backend_game.piece_placement(row, col)
+                # print(x, "\n")
+                if x == 1:
+                    self.draw_piece(self.backend_game.get_turn(),current_cell * self.cell_size)
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/BoardGUI.py:48]: formatting nit: should be a space between parameters.

File: frontend/BoardGUI.py (hunk 7 above)
Comment (raspberrymilkyway) [reply frontend/BoardGUI.py:49]: Hmm, yeah, we could just make another function to draw pieces and swap turns that can be called in the GameGUI when a valid cell is returned, like how the board is drawn.

Comment (raspberrymilkyway) [reply frontend/BoardGUI.py:49]: That might make programming this section more of a pain later, though -- piece placement, movement, flying, and removing pieces are all determined by clicking. I guess we could add a get_phase function in BoardGUI and call it later to determine exactly what should be happening with the returned cell.

File: backend/GameManager.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    def placement_complete(self):
+        return self.pieces.all_pieces_placed()
+
+    def piece_placement(self, row, column):
+        if (row, column) in self.current_moves:
+            if self.board.check_position(row, column) != "empty":
+                print(self.board.check_position(row, column).lower())
+                return "GameManagerError -- position not empty"
+            
+            flag = 0
```
Comment (EXC3ll3NTrhyTHM) [inline backend/GameManager.py:34]: We could name this isPiecedPlaced or something similar to capture what the variable represents

Comment (EXC3ll3NTrhyTHM) [reply backend/GameManager.py:34]: May even want to throw exception instead of return `False` in the set_piece functions then you can try-except it here, then we could do something on screen to communicate that it was invalid position without changing how the set_piece function returns values.

File: backend/Piece.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
+from enum import Enum
+
+
+class Turn(Enum):
```
Comment (EXC3ll3NTrhyTHM) [inline backend/Piece.py:4]: I feel like this could be passed into the Piece class from a different package but Im not sure about if that architecture makes more sense.

Comment (EXC3ll3NTrhyTHM) [reply backend/Piece.py:4]: To me Turn is more cohesive in the GameManager package.

File: backend/Piece.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            return Turn.WHITE
+        else:
+            return Turn.BLACK
+
+class Piece:
+    def __init__(self, color):
+        self.color = color
+        self.position = (-1, -1)
+    
+    def get_turn(self):
```
Comment (EXC3ll3NTrhyTHM) [inline backend/Piece.py:20]: This doesnt feel like it belongs in this class. Unless its `get_color`. I think this should be called `get_color`.

File: backend/Piece.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        return self.color
+    def get_position(self):
+        return self.position
+    
+    def set_turn(self, color):
```
Comment (EXC3ll3NTrhyTHM) [inline backend/Piece.py:25]: This should be `set_color`. Turn is a concept that the GameManager worries about and involves switching. Color for a given piece is static.

File: frontend/BoardGUI.py (hunk 7 above)
Comment (EXC3ll3NTrhyTHM) [reply frontend/BoardGUI.py:49]: Adding a `get_phase` is a good idea.

File: backend/Piece.py (hunk 10 above)
Comment (raspberrymilkyway) [reply backend/Piece.py:4]: I don't think Turn should be in the GameManager, even if it's more cohesive, because I'm calling it in the Pieces class. It can be in its own separate class, but we should not import anything from GameManager into the Piece.py file.

I guess I can redo how Piece colors are stored, if we'd prefer that -- Turn would be fine to move to GameManager then.

File: backend/Piece.py (hunk 11 above)
Comment (raspberrymilkyway) [reply backend/Piece.py:20]: It was supposed to be -- I swapped the name of the variable after the I made the Turn class. Must've forgotten to swap the function names.

File: backend/Piece.py (hunk 12 above)
Comment (raspberrymilkyway) [reply backend/Piece.py:25]: See previous.

File: backend/GameManager.py (hunk 9 above)
Comment (raspberrymilkyway) [reply backend/GameManager.py:34]: That was something I was going to ask about, but we didn't meet with everyone after I did Pieces and honestly I didn't think it mattered that much for this first sprint: we never discussed how we wanted to do error handling. The way I have it set up now is easy to find and change, so we can absolutely do that, but I do think we should discuss it at some point.
